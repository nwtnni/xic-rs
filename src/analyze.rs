mod anticipated_expressions;
mod available_expressions;
mod call_graph;
mod conditional_constant_propagation;
mod constant_propagation;
mod copy_propagation;
mod dot;
mod earliest;
mod latest;
mod live_ranges;
mod live_variables;
mod postponable_expressions;
mod used_expressions;
mod valid_allocation;

pub use anticipated_expressions::AnticipatedExpressions;
pub use available_expressions::AvailableExpressions;
pub use call_graph::CallGraph;
pub use conditional_constant_propagation::ConditionalConstantPropagation;
pub use conditional_constant_propagation::Constant;
pub use conditional_constant_propagation::Reachable;
pub use constant_propagation::ConstantPropagation;
pub use copy_propagation::CopyPropagation;
pub use dot::display;
pub use earliest::Earliest;
pub use latest::Latest;
pub use live_ranges::LiveRanges;
pub use live_ranges::Range;
pub use live_variables::LiveVariables;
pub use postponable_expressions::PostponableExpressions;
pub use used_expressions::UsedExpressions;

use std::collections::VecDeque;

use petgraph::Direction;

use crate::cfg;
use crate::cfg::Cfg;
use crate::cfg::Function;
use crate::data::operand::Label;
use crate::util;
use crate::Map;
use crate::Set;

pub trait Analysis<T: Function>: Sized {
    const BACKWARD: bool;

    type Data: Clone + Eq;

    fn default(&self) -> Self::Data;

    fn default_with_metadata(&self, _label: &Label) -> Self::Data {
        self.default()
    }

    fn transfer(&self, statement: &T::Statement, output: &mut Self::Data);

    fn transfer_with_metadata(
        &self,
        _label: &Label,
        _index: usize,
        statement: &T::Statement,
        output: &mut Self::Data,
    ) {
        self.transfer(statement, output);
    }

    fn merge<'a, I>(&self, outputs: I, input: &mut Self::Data)
    where
        I: Iterator<Item = Option<&'a Self::Data>>,
        Self::Data: 'a;

    fn merge_with_metadata<'a, I>(&self, outputs: I, input: &mut Self::Data)
    where
        I: Iterator<Item = (&'a cfg::Edge, Option<&'a Self::Data>)>,
        Self::Data: 'a,
    {
        self.merge(outputs.map(|(_, data)| data), input)
    }
}

pub struct Solution<A: Analysis<T>, T: Function> {
    pub analysis: A,
    pub inputs: Map<Label, A::Data>,
    pub outputs: Map<Label, A::Data>,
}

pub fn analyze_default<A, T>(cfg: &Cfg<T>) -> Solution<A, T>
where
    A: Analysis<T> + Default,
    T: Function,
{
    analyze(Default::default(), cfg)
}

pub fn analyze<A: Analysis<T>, T: Function>(analysis: A, cfg: &Cfg<T>) -> Solution<A, T> {
    log::info!(
        "[{}] Running {} analysis on {}...",
        std::any::type_name::<Cfg<T>>(),
        std::any::type_name::<A>(),
        cfg.name(),
    );
    util::time!(
        "[{}] Done running {} analysis on {}",
        std::any::type_name::<Cfg<T>>(),
        std::any::type_name::<A>(),
        cfg.name(),
    );

    let strongly_connected_components = cfg.strongly_connected_components(A::BACKWARD);
    let (predecessors, successors) = match A::BACKWARD {
        true => (Direction::Outgoing, Direction::Incoming),
        false => (Direction::Incoming, Direction::Outgoing),
    };

    let mut inputs = Map::<Label, A::Data>::default();
    let mut outputs = Map::<Label, A::Data>::default();
    let mut worklist = VecDeque::new();
    let mut workset = Set::default();

    // Topological order
    for component in strongly_connected_components.into_iter().rev() {
        // Reverse postorder
        worklist.extend(component.iter().copied().rev());
        workset.extend(component.iter().copied().rev());

        let component = component.into_iter().collect::<Set<_>>();

        while let Some(label) = worklist.pop_front() {
            workset.remove(&label);

            let input = inputs
                .entry(label)
                .or_insert_with(|| analysis.default_with_metadata(&label));

            analysis.merge_with_metadata(
                cfg.edges_directed(predecessors, &label)
                    .map(|(predecessor, edge)| (edge, outputs.get(&predecessor))),
                input,
            );

            let mut output = input.clone();

            if A::BACKWARD {
                for (index, statement) in cfg[&label].iter().enumerate().rev() {
                    analysis.transfer_with_metadata(&label, index, statement, &mut output);
                }
            } else {
                for (index, statement) in cfg[&label].iter().enumerate() {
                    analysis.transfer_with_metadata(&label, index, statement, &mut output);
                }
            }

            match outputs.get_mut(&label) {
                Some(existing) if *existing == output => continue,
                Some(existing) => *existing = output,
                None => {
                    outputs.insert(label, output);
                }
            };

            cfg.neighbors(successors, &label)
                .filter(|successor| component.contains(successor))
                .filter(|successor| workset.insert(*successor))
                .for_each(|successor| worklist.push_back(successor));
        }
    }

    Solution {
        analysis,
        inputs,
        outputs,
    }
}
