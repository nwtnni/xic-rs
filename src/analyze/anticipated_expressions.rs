use std::collections::BTreeSet;

use crate::analyze::Analysis;
use crate::data::lir;

pub struct AnticipatedExpressions;

impl<T: lir::Target> Analysis<lir::Function<T>> for AnticipatedExpressions {
    const BACKWARD: bool = true;

    type Data = BTreeSet<lir::Expression>;

    fn new() -> Self {
        Self
    }

    fn default(&self) -> Self::Data {
        BTreeSet::new()
    }

    fn transfer(&self, statement: &lir::Statement<T>, output: &mut Self::Data) {
        match statement {
            lir::Statement::Jump(_) | lir::Statement::Label(_) => (),
            lir::Statement::CJump {
                condition: _,
                left,
                right,
                r#true: _,
                r#false: _,
            } => {
                output.insert(left.clone());
                output.insert(right.clone());
            }
            lir::Statement::Call(function, arguments, returns) => {
                for r#return in 0..*returns {
                    Self::remove(output, &lir::Expression::Return(r#return));
                }

                output.extend(arguments.iter().cloned());
                output.insert(function.clone());
            }
            lir::Statement::Move {
                destination,
                source,
            } => {
                Self::remove(output, destination);
                output.insert(source.clone());
            }
            lir::Statement::Return(returns) => {
                output.extend(returns.iter().cloned());
            }
        }
    }

    fn merge<'a, I>(&self, mut outputs: I, input: &mut Self::Data)
    where
        I: Iterator<Item = Option<&'a Self::Data>>,
        Self::Data: 'a,
    {
        input.clear();
        input.extend(outputs.next().into_iter().flatten().flatten().cloned());

        let mut stack = Vec::new();

        for output in outputs.flatten() {
            stack.extend(
                input
                    .iter()
                    .filter(|expression| !output.contains(expression))
                    // Could be avoided with https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.drain_filter
                    .cloned(),
            );

            stack.drain(..).for_each(|kill| Self::remove(input, &kill));
        }
    }
}

impl AnticipatedExpressions {
    pub(super) fn remove(output: &mut BTreeSet<lir::Expression>, kill: &lir::Expression) {
        output.remove(kill);

        let mut stack = vec![kill.clone()];

        while let Some(killed) = stack.pop() {
            output.retain(|kill| {
                if !Self::contains(kill, &killed) {
                    return true;
                }

                stack.push(kill.clone());
                false
            })
        }
    }

    fn contains(expression: &lir::Expression, killed: &lir::Expression) -> bool {
        match expression {
            lir::Expression::Argument(_)
            | lir::Expression::Return(_)
            | lir::Expression::Immediate(_)
            | lir::Expression::Temporary(_) => expression == killed,
            lir::Expression::Memory(address) => Self::contains(&*address, killed),
            lir::Expression::Binary(_, left, right) => {
                Self::contains(&*left, killed) || Self::contains(&*right, killed)
            }
        }
    }
}
