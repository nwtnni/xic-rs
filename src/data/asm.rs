use crate::data::ir;
use crate::data::operand;

pub type Unit<T> = ir::Unit<Function<T>>;
pub type Function<T> = ir::Function<Vec<Assembly<T>>>;

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Assembly<T> {
    Binary(Binary, operand::Binary<T>),
    Unary(Unary, operand::Unary<T>),
    Nullary(Nullary),
    Label(operand::Label),
    Directive(Directive),
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Directive {
    Intel,
    Align(usize),
    Local(operand::Label),
    Global(operand::Label),
    Quad(Vec<i64>),
    Data,
    Text,
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Binary {
    Add,
    Sub,
    And,
    Or,
    Xor,
    Cmp,
    Mov,
    Lea,
}

impl From<ir::Binary> for Binary {
    fn from(binary: ir::Binary) -> Self {
        match binary {
            ir::Binary::Add => Binary::Add,
            ir::Binary::Sub => Binary::Sub,
            ir::Binary::Xor => Binary::Xor,
            ir::Binary::And => Binary::And,
            ir::Binary::Or => Binary::Or,
            _ => panic!("[INTERNAL ERROR]: converting unsupported IR operator"),
        }
    }
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Unary {
    Neg,
    Push,
    Pop,
    Call { arguments: usize, returns: usize },
    Mul,
    Div(Division),
    Jmp,
    Jcc(Condition),
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Nullary {
    Cqo,
    Ret,
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Condition {
    L,
    Le,
    Ge,
    G,
    Ne,
    E,
}

impl From<ir::Condition> for Condition {
    fn from(condition: ir::Condition) -> Self {
        match condition {
            ir::Condition::Lt => Condition::L,
            ir::Condition::Le => Condition::Le,
            ir::Condition::Ge => Condition::Ge,
            ir::Condition::Gt => Condition::G,
            ir::Condition::Ne => Condition::Ne,
            ir::Condition::Eq => Condition::E,
        }
    }
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Division {
    Quotient,
    Remainder,
}
