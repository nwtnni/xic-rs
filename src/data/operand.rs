use std::sync::atomic::{AtomicUsize, Ordering};

use crate::data::symbol;

static LABELS: AtomicUsize = AtomicUsize::new(0);
static TEMPS: AtomicUsize = AtomicUsize::new(0);

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub enum Immediate {
    Constant(i64),
    Label(Label),
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub enum Label {
    Fixed(symbol::Symbol),
    Fresh(symbol::Symbol, usize),
}

impl Label {
    pub fn fresh(label: &'static str) -> Self {
        let index = LABELS.fetch_add(1, Ordering::SeqCst);
        let symbol = symbol::intern_static(label);
        Label::Fresh(symbol, index)
    }
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub enum Temporary {
    Register(Register),
    Argument(usize),
    Return(usize),
    Fresh(symbol::Symbol, usize),
}

impl Temporary {
    pub fn fresh(label: &'static str) -> Self {
        let index = TEMPS.fetch_add(1, Ordering::SeqCst);
        let symbol = symbol::intern_static(label);
        Temporary::Fresh(symbol, index)
    }
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub enum Register {
    Rax,
    Rbx,
    Rcx,
    Rdx,
    Rbp,
    Rsp,
    Rsi,
    Rdi,
    R8,
    R9,
    R10,
    R11,
    R12,
    R13,
    R14,
    R15,
}

pub trait Operand: Copy + Eq + std::hash::Hash + std::fmt::Debug + PartialOrd + Ord {}
impl Operand for Temporary {}
impl Operand for Register {}

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub enum Memory<T: Operand> {
    R(T),
    RO(T, i32),
}
