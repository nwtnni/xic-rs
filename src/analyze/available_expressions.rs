use std::collections::BTreeSet;

use crate::analyze::analyze;
use crate::analyze::Analysis;
use crate::analyze::AnticipatedExpressions;
use crate::analyze::Solution;
use crate::data::lir;
use crate::data::operand::Label;

pub struct AvailableExpressions<T: lir::Target> {
    anticipated: Solution<AnticipatedExpressions, lir::Function<T>>,
}

impl<T: lir::Target> Analysis<lir::Function<T>> for AvailableExpressions<T> {
    const BACKWARD: bool = false;

    type Data = BTreeSet<lir::Expression>;

    fn new() -> Self {
        unreachable!()
    }

    fn new_with_metadata(cfg: &crate::cfg::Cfg<lir::Function<T>>) -> Self {
        Self {
            anticipated: analyze(cfg),
        }
    }

    fn default(&self) -> Self::Data {
        unreachable!()
    }

    fn default_with_metadata(&self, label: &Label) -> Self::Data {
        self.anticipated.outputs[label].clone()
    }

    fn transfer(&self, statement: &lir::Statement<T>, output: &mut Self::Data) {
        match statement {
            lir::Statement::Jump(_)
            | lir::Statement::CJump { .. }
            | lir::Statement::Label(_)
            | lir::Statement::Return(_) => (),
            lir::Statement::Call(_, _, returns) => {
                for r#return in 0..*returns {
                    AnticipatedExpressions::remove(output, &lir::Expression::Return(r#return));
                }
            }
            lir::Statement::Move {
                destination,
                source: _,
            } => {
                AnticipatedExpressions::remove(output, destination);
            }
        }
    }

    fn merge<'a, I>(&self, outputs: I, input: &mut Self::Data)
    where
        I: Iterator<Item = Option<&'a Self::Data>>,
        Self::Data: 'a,
    {
        <AnticipatedExpressions as Analysis<lir::Function<T>>>::merge(
            &self.anticipated.analysis,
            outputs,
            input,
        );
    }
}
