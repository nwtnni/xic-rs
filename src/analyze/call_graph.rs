use petgraph::graphmap::DiGraphMap;
use petgraph::visit::Walker as _;
use petgraph::Direction;

use crate::data::lir;
use crate::data::operand::{Immediate, Label};
use crate::data::symbol::Symbol;

pub struct CallGraph(DiGraphMap<Symbol, ()>);

impl CallGraph {
    pub fn new<T: lir::Target>(lir: &lir::Unit<T>) -> Self {
        let mut graph = lir
            .functions
            .iter()
            .flat_map(|(caller, function)| {
                function
                    .statements
                    .iter()
                    .map(move |statement| (caller, statement))
            })
            .filter_map(|(caller, statement)| match statement {
                lir::Statement::Call(
                    lir::Expression::Immediate(Immediate::Label(Label::Fixed(callee))),
                    _,
                    _,
                ) => Some((*caller, *callee)),
                _ => None,
            })
            .collect::<DiGraphMap<_, _>>();

        for name in lir.functions.keys() {
            graph.add_node(*name);
        }

        Self(graph)
    }

    pub fn is_recursive(&self, name: &Symbol) -> bool {
        self.0.contains_edge(*name, *name)
    }

    pub fn is_leaf(&self, name: &Symbol) -> bool {
        self.0
            .neighbors_directed(*name, Direction::Outgoing)
            .count()
            == 0
    }

    pub fn postorder(&self, start: &Symbol) -> impl Iterator<Item = Symbol> + '_ {
        petgraph::visit::DfsPostOrder::new(&self.0, *start).iter(&self.0)
    }
}
