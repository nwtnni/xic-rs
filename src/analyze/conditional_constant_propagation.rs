use crate::analyze::Analysis;
use crate::cfg::Cfg;
use crate::cfg::Edge;
use crate::data::ir;
use crate::data::lir;
use crate::data::operand::Immediate;
use crate::data::operand::Label;
use crate::data::operand::Temporary;
use crate::optimize;
use crate::Map;

pub struct ConditionalConstantPropagation {
    enter: Label,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Data {
    pub(crate) reachable: Reachable,
    pub(crate) constants: Map<Temporary, Constant>,
}

#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum Reachable {
    Linear(bool),
    Branch(bool, bool),
}

#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum Constant {
    Defined(Immediate),
    Overdefined,
}

impl<T: lir::Target> Analysis<lir::Function<T>> for ConditionalConstantPropagation {
    const BACKWARD: bool = false;

    type Data = Data;

    fn new() -> Self {
        unreachable!()
    }

    fn new_with_metadata(cfg: &Cfg<lir::Function<T>>) -> Self {
        ConditionalConstantPropagation {
            enter: *cfg.enter(),
        }
    }

    fn default(&self) -> Self::Data {
        unreachable!()
    }

    fn default_with_metadata(&self, label: &Label) -> Self::Data {
        Data {
            reachable: Reachable::Linear(*label == self.enter),
            constants: Map::default(),
        }
    }

    fn transfer(&self, statement: &lir::Statement<T>, output: &mut Self::Data) {
        match output.reachable {
            Reachable::Branch(_, _) => unreachable!(),
            Reachable::Linear(true) => (),
            Reachable::Linear(false) => {
                output.reachable = match statement {
                    lir::Statement::CJump { .. } => Reachable::Branch(false, false),
                    lir::Statement::Jump(_)
                    | lir::Statement::Call(_, _, _)
                    | lir::Statement::Label(_)
                    | lir::Statement::Move { .. }
                    | lir::Statement::Return(_) => Reachable::Linear(false),
                };
                return;
            }
        }

        match statement {
            lir::Statement::Jump(_)
            | lir::Statement::Label(_)
            | lir::Statement::Call(_, _, _)
            | lir::Statement::Move {
                destination: lir::Expression::Memory(_),
                source: _,
            }
            // Note: normally, we would mark anything after a return as unreachable,
            // but we desugar returns into move + jump statements when we tile assembly.
            | lir::Statement::Return(_) => {}
            lir::Statement::CJump {
                condition,
                left,
                right,
                r#true: _,
                r#false: _,
            } => {
                // FIXME: collect symbolic constraints based on branch
                output.reachable = match output.evaluate_condition(condition, left, right) {
                    None => Reachable::Branch(true, true),
                    Some(true) => Reachable::Branch(true, false),
                    Some(false) => Reachable::Branch(false, true),
                }
            }
            lir::Statement::Move {
                destination: lir::Expression::Temporary(temporary),
                source,
            } => {
                if let Some(immediate) = output.evaluate_expression(source) {
                    output
                        .constants
                        .entry(*temporary)
                        .and_modify(|constant| match constant {
                            Constant::Overdefined => (),
                            Constant::Defined(_) => *constant = Constant::Defined(immediate),
                        })
                        .or_insert(Constant::Defined(immediate));
                }
            }
            lir::Statement::Move { .. } => unreachable!(),
        }
    }

    fn merge<'a, I>(&self, _: I, _: &mut Self::Data)
    where
        I: Iterator<Item = Option<&'a Self::Data>>,
        Self::Data: 'a,
    {
        unreachable!()
    }

    fn merge_with_metadata<'a, I>(&self, outputs: I, input: &mut Self::Data)
    where
        I: Iterator<Item = (&'a Edge, Option<&'a Self::Data>)>,
        Self::Data: 'a,
    {
        input.constants.clear();

        let reachable = match &mut input.reachable {
            Reachable::Linear(reachable) => reachable,
            Reachable::Branch(_, _) => unreachable!(),
        };

        for (_, output) in outputs
            .filter_map(|(edge, output)| output.map(move |output| (edge, output)))
            // Ignore merge information from unreachable predecessors
            .filter(|(edge, output)| {
                matches!(
                    (edge, output.reachable),
                    (Edge::Unconditional, Reachable::Linear(true))
                        | (Edge::Conditional(true), Reachable::Branch(true, _))
                        | (Edge::Conditional(false), Reachable::Branch(_, true))
                )
            })
        {
            *reachable = true;

            // Defined in `input` and `output`
            input.constants.retain(
                |temporary, old| match (old, output.constants.get(temporary)) {
                    (Constant::Defined(_), None) => true,
                    (Constant::Defined(old), Some(Constant::Defined(new))) if old == new => true,
                    (old, _) => {
                        *old = Constant::Overdefined;
                        true
                    }
                },
            );

            // Defined in `output`
            for (temporary, constant) in &output.constants {
                input.constants.entry(*temporary).or_insert(*constant);
            }
        }
    }
}

impl Data {
    pub fn evaluate_condition(
        &self,
        condition: &ir::Condition,
        left: &lir::Expression,
        right: &lir::Expression,
    ) -> Option<bool> {
        let (left, right) = match (
            self.evaluate_expression(left)?,
            self.evaluate_expression(right)?,
        ) {
            (Immediate::Integer(left), Immediate::Integer(right)) => (left, right),
            _ => return None,
        };

        Some(optimize::fold_condition(*condition, left, right))
    }

    pub fn evaluate_expression(&self, expression: &lir::Expression) -> Option<Immediate> {
        match expression {
            lir::Expression::Argument(_)
            | lir::Expression::Return(_)
            // Conservatively mark all memory as unknown for now.
            | lir::Expression::Memory(_) => None,
            lir::Expression::Immediate(immediate) => Some(*immediate),
            lir::Expression::Temporary(temporary) => match self.constants.get(temporary)? {
                Constant::Defined(immediate) => Some(*immediate),
                Constant::Overdefined => None,
            },
            lir::Expression::Binary(binary, left, right) => {
                // FIXME: handle labels with offsets at some point
                let (left, right) = match (self.evaluate_expression(left)?, self.evaluate_expression(right)?) {
                    (Immediate::Integer(left), Immediate::Integer(right)) => (left, right),
                    _ => return None,
                };

                // FIXME: handle division by 0, algebraic identities
                Some(Immediate::Integer(optimize::fold_binary(*binary, left, right)))
            }
        }
    }
}
