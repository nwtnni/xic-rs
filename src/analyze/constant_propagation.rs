use crate::abi;
use crate::analyze::Analysis;
use crate::data::asm;
use crate::data::operand;
use crate::data::operand::Immediate;
use crate::data::operand::Register;
use crate::data::operand::Temporary;
use crate::util;
use crate::Map;

pub struct ConstantPropagation;

impl Analysis<asm::Function<Temporary>> for ConstantPropagation {
    const BACKWARD: bool = false;

    type Data = Map<Temporary, Immediate>;

    fn new() -> Self {
        ConstantPropagation
    }

    fn default(&self) -> Self::Data {
        Map::default()
    }

    fn transfer(&self, statement: &asm::Statement<Temporary>, output: &mut Self::Data) {
        use asm::Binary::*;
        use asm::Nullary::*;
        use asm::Unary::*;

        match statement {
            asm::Statement::Binary(Mov, operands) => match operands {
                operand::Binary::RI {
                    destination,
                    source,
                } => {
                    output.insert(*destination, *source);
                }
                operand::Binary::RR {
                    destination,
                    source,
                } => {
                    if let Some(immediate) = output.get(source).copied() {
                        output.insert(*destination, immediate);
                    } else {
                        output.remove(destination);
                    }
                }
                operand::Binary::RM { destination, .. } => {
                    output.remove(destination);
                }
                operand::Binary::MR { .. } | operand::Binary::MI { .. } => (),
            },
            asm::Statement::Binary(Lea, operands) => {
                if let util::Or::L(temporary) = operands.destination() {
                    output.remove(&temporary);
                }
            }
            asm::Statement::Binary(binary @ (Add | Sub | Shl | Mul | And | Or | Xor), operands) => {
                match operands {
                    operand::Binary::RI {
                        destination,
                        source,
                    } => {
                        if let Some(immediate) = output
                            .get(destination)
                            .and_then(|immediate| fold_binary(binary, immediate, source))
                        {
                            output.insert(*destination, immediate);
                        } else {
                            output.remove(destination);
                        }
                    }
                    operand::Binary::RM { destination, .. } => {
                        output.remove(destination);
                    }
                    operand::Binary::RR {
                        destination,
                        source,
                    } => {
                        if let Some(immediate) =
                            output.get(destination).zip(output.get(source)).and_then(
                                |(destination, source)| fold_binary(binary, destination, source),
                            )
                        {
                            output.insert(*destination, immediate);
                        } else {
                            output.remove(destination);
                        }
                    }
                    operand::Binary::MI { .. } | operand::Binary::MR { .. } => (),
                }
            }
            asm::Statement::Binary(Cmp, _) => (),
            asm::Statement::Unary(unary @ (Hul | Div | Mod), operand) => {
                let immediate = match operand {
                    operand::Unary::I(immediate) => Some(immediate),
                    operand::Unary::R(temporary) => output.get(temporary),
                    operand::Unary::M(_) => None,
                };

                if let Some((rax, rdx)) = output
                    .get(&Temporary::Register(Register::Rax))
                    .zip(immediate)
                    .and_then(|(destination, source)| fold_unary(unary, destination, source))
                {
                    output.insert(Temporary::Register(Register::Rax), rax);
                    output.insert(Temporary::Register(Register::Rdx), rdx);
                } else {
                    output.remove(&Temporary::Register(Register::Rax));
                    output.remove(&Temporary::Register(Register::Rdx));
                }
            }
            asm::Statement::Unary(Neg, operand::Unary::R(temporary)) => {
                if let Some(Immediate::Integer(integer)) = output.get(temporary).copied() {
                    output.insert(*temporary, Immediate::Integer(-integer));
                } else {
                    output.remove(temporary);
                }
            }
            asm::Statement::Unary(Neg, operand::Unary::I(_) | operand::Unary::M(_)) => (),
            asm::Statement::Unary(
                Call {
                    arguments: _,
                    returns,
                },
                _,
            ) => {
                for register in abi::CALLER_SAVED
                    .iter()
                    .chain(abi::RETURN.iter().take(*returns))
                {
                    output.remove(&Temporary::Register(*register));
                }
            }

            asm::Statement::Nullary(Cqo) => {
                output.insert(Temporary::Register(Register::Rdx), Immediate::Integer(0));
            }
            asm::Statement::Nullary(Nop | Ret(_))
            | asm::Statement::Label(_)
            | asm::Statement::Jmp(_)
            | asm::Statement::Jcc(_, _) => (),
        }
    }

    fn merge<'a, I>(&self, mut outputs: I, input: &mut Self::Data)
    where
        I: Iterator<Item = Option<&'a Self::Data>>,
        Self::Data: 'a,
    {
        input.clear();
        input.extend(outputs.next().into_iter().flatten().flatten());

        for output in outputs {
            let output = match output {
                Some(output) => output,
                None => {
                    input.clear();
                    return;
                }
            };

            input.retain(|temporary, old| match output.get(temporary) {
                Some(new) if old == new => true,
                Some(_) | None => false,
            });
        }
    }
}

fn fold_binary(
    binary: &asm::Binary,
    destination: &Immediate,
    source: &Immediate,
) -> Option<Immediate> {
    let (destination, source) = match (destination, source) {
        (Immediate::Integer(destination), Immediate::Integer(source)) => (*destination, *source),
        (_, _) => return None,
    };

    let integer = match binary {
        asm::Binary::Add => destination.wrapping_add(source),
        asm::Binary::Sub => destination.wrapping_sub(source),
        asm::Binary::Shl => destination << source,
        asm::Binary::Mul => destination.wrapping_mul(source),
        asm::Binary::And => destination & source,
        asm::Binary::Or => destination | source,
        asm::Binary::Xor => destination ^ source,
        asm::Binary::Cmp | asm::Binary::Mov | asm::Binary::Lea => return None,
    };

    Some(Immediate::Integer(integer))
}

fn fold_unary(
    unary: &asm::Unary,
    destination: &Immediate,
    source: &Immediate,
) -> Option<(Immediate, Immediate)> {
    let (destination, source) = match (destination, source) {
        (Immediate::Integer(destination), Immediate::Integer(source)) => (*destination, *source),
        (_, _) => return None,
    };

    let (rax, rdx) = match unary {
        asm::Unary::Neg | asm::Unary::Call { .. } => return None,
        asm::Unary::Hul => (
            destination.wrapping_mul(source),
            (((destination as i128) * (source as i128)) >> 64) as i64,
        ),
        asm::Unary::Div | asm::Unary::Mod if source == 0 => return None,
        asm::Unary::Div | asm::Unary::Mod => (destination / source, destination % source),
    };

    Some((Immediate::Integer(rax), Immediate::Integer(rdx)))
}
