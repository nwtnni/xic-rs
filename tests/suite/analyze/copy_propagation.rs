use xic::api::analyze::analyze;
use xic::api::analyze::display;
use xic::api::analyze::CopyPropagation;
use xic::data::asm::Function;
use xic::data::operand::Temporary;

macro_rules! copy_propagation {
    ($function:ident $($tt:tt)*) => {
        #[test]
        fn $function() {
            let function = assembly!($function $($tt)*);
            let copy_propagation = copy_propagation(function);
            insta::assert_display_snapshot!(copy_propagation);
        }
    }
}

fn copy_propagation(function: Function<Temporary>) -> String {
    let cfg = xic::api::construct_cfg(function);
    let copy_propagation = analyze::<CopyPropagation, _>(&cfg);
    let annotated_cfg = super::super::graph(display(&copy_propagation, &cfg));
    annotated_cfg
}

copy_propagation! {
    redundant_copies: 0 -> 0;
    temporaries: a, b, c, d, e;
    (mov b, a)
    (mov c, b)
    (mov d, c)
    (mov e, d)
}

copy_propagation! {
    overwritten_tree: 0 -> 0;
    temporaries: a, b, c, d;
    (mov a, b)
    (mov c, a)
    (mov d, a)
    (mov b, 1)
}

copy_propagation! {
    clobbered_across_call: 0 -> 0;
    temporaries: a;
    labels: black_box;
    (mov rax, a)
    (mov rcx, a)
    (mov rdx, a)
    (mov rsi, a)
    (mov rdi, a)
    (mov r8, a)
    (mov r9, a)
    (mov r10, a)
    (mov r11, a)
    (call<0, 0> black_box)
}

copy_propagation! {
    clobbered_across_mul: 0 -> 0;
    temporaries: a;
    (mov rax, a)
    (mov rdx, a)
    (imul 1)
}

copy_propagation! {
    clobbered_across_hul: 0 -> 0;
    temporaries: a;
    (mov rax, a)
    (mov rdx, a)
    (ihul 1)
}

copy_propagation! {
    clobbered_across_div: 0 -> 0;
    temporaries: a;
    (mov rax, a)
    (mov rdx, a)
    (idiv 1)
}

copy_propagation! {
    clobbered_across_mod: 0 -> 0;
    temporaries: a;
    (mov rax, a)
    (mov rdx, a)
    (imod 1)
}

copy_propagation! {
    clobbered_across_cqo: 0 -> 0;
    temporaries: a;
    (mov rax, a)
    (mov rdx, a)
    (cqo)
}

copy_propagation! {
    partially_overwritten_tree: 0 -> 0;
    temporaries: a, b, c, d, e;
    (mov a, b)
    (mov d, a)
    (mov c, b)
    (mov e, c)
    (mov a, 1)
}

copy_propagation! {
    defined_twice_different: 0 -> 0;
    temporaries: a, b, c, d;
    labels: exit, branch, fallthrough;
    (cmp a, b)
    (mov d, a)
    (jne branch)
    (fallthrough:)
    (mov a, b)
    (jmp exit)
    (branch:)
    (mov a, c)
    (jmp exit)
}

copy_propagation! {
    defined_twice_identical: 0 -> 0;
    temporaries: a, b, c;
    labels: exit, branch, fallthrough;
    (cmp a, b)
    (mov c, a)
    (jne branch)
    (fallthrough:)
    (mov a, b)
    (jmp exit)
    (branch:)
    (mov a, b)
    (jmp exit)
}
