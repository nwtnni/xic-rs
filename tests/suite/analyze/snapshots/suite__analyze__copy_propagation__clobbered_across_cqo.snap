---
source: tests/suite/analyze/copy_propagation.rs
assertion_line: 91
expression: copy_propagation
---

clobbered_across_cqo

+--------------------+
|       enter:       |
| {}                 |
| mov rax, _a        |
| {rax: _a}          |
| mov rdx, _a        |
| {rax: _a, rdx: _a} |
| cqo                |
| {}                 |
| jmp exit           |
| {}                 |
+--------------------+
  |
  |
  v
+--------------------+
|       exit:        |
| {}                 |
| ret                |
| {}                 |
+--------------------+

